File: 414-third-maximum-number/414-third-maximum-number.py
class Solution:
    def thirdMax(self, nums: List[int]) -> int:
        threeLargest = [None, None, None]
        nums = list(set(nums))
        
        for num in nums:
            self.updateLargest(threeLargest, num)
        
        return threeLargest[0] if threeLargest[0] is not None else threeLargest[2]
        
    def updateLargest(self, array, num):
        if array[2] is None or array[2]<= num:
            self.shiftAndUpdate(array, num, 2) 
        elif array[1] is None or array[1]<= num:
            self.shiftAndUpdate(array, num, 1)
        elif array[0] is None or array[0]<= num:
            self.shiftAndUpdate(array, num, 0)
    
    def shiftAndUpdate(self, array, num, idx):
        for i in range(idx+1):
            if i==idx:
                array[i]=num
            else:
                array[i]=array[i+1]

File: 414-third-maximum-number/414-third-maximum-number.java
class Solution {
    public int thirdMax(int[] nums) {
        Integer Max1=null;
        Integer Max2=null;
        Integer Max3=null;
        
        for(Integer n:nums){
            if(n.equals(Max1) || n.equals(Max2) || n.equals(Max3)) continue;
            if(Max1==null || n>Max1){
                Max3=Max2;
                Max2=Max1;
                Max1=n;
            } else if(Max2==null || n>Max2){
                Max3=Max2;
                Max2=n;
            } else if(Max3==null || n>Max3){
                Max3=n;
            }
        }
        
        return Max3==null? Max1:Max3;
        
    }
}